### PR TITLE
quantile filter support

### LINKED ISSUE #1543: Support quantile filter for sensor values
**Describe the problem you have/What new integration you would like**

I have sonar distance meter which has dispersion in measured values but picking quantile gives good results.

**Please describe your use case for this integration and alternatives you've tried:**

I'd like to have quantile function have available directly in the esphome. Currently I use prometheus to aggragate data but this is another system in the middle. It will be nicer to have this function right in esphome.
Parameter could be numper of frames and the quantile or percentile value. Then quantile aggrgated from the count of frames will be provided.

**Additional context**

### PR DESCRIPTION
**Pull request by konikvranik**
# What does this implement/fix? 

Implemented quantile filter for sensor.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes https://github.com/esphome/feature-requests/issues/1543

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1715

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [x] ESP8266

## Example entry for `config.yaml`:

```yaml
  - platform: template
...
    filters:
      - quantile:
          window_size: 120
          send_every: 12
          send_first_at: 6
          quantile: .7
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Review by jesserockz — APPROVED** [review]
Could you add this to `tests/test3.yaml` please?

File: esphome/components/sensor/filter.cpp [hunk 1]
```diff
@@ -66,6 +67,41 @@ optional<float> MedianFilter::new_value(float value) {
...
+  if (++this->send_at_ >= this->send_every_) {
+    this->send_at_ = 0;
+
+    float result = 0.0f;
+    if (!this->queue_.empty()) {
+      std::deque<float> quantile_queue = this->queue_;
+      sort(quantile_queue.begin(), quantile_queue.end());
+
+      size_t queue_size = quantile_queue.size();
+      size_t position = ceill(queue_size * this->quantile_) - 1;
```
Comment (oxan) [inline esphome/components/sensor/filter.cpp]: No reason to use `long double` (which utterly kills performance) here.
```suggestion
      size_t position = ceilf(queue_size * this->quantile_) - 1;
```

File: esphome/components/sensor/filter.h [hunk 2]
```diff
@@ -42,6 +42,37 @@ class Filter {
   Sensor *parent_{nullptr};
 };
 
+/** Simple quantile filter.
+ *
+ * Takes the median of the last <send_every> values and pushes it out every <send_every>.
```
Comment (oxan) [inline esphome/components/sensor/filter.h]: Copy/paste error.

File: esphome/components/sensor/filter.h [hunk 3]
```diff
@@ -42,6 +42,37 @@ class Filter {
...
+ */
+class QuantileFilter : public Filter {
+ public:
+  /** Construct a QuantileFilter.
+   *
+   * @param window_size The number of values that should be used in median calculation.
```
Comment (oxan) [inline esphome/components/sensor/filter.h]: Copy/paste error.

**Review by oxan — CHANGES_REQUESTED** [review]

**Review by oxan — APPROVED** [review]

**Pull request merged**